components/cloud/aws/src/s3.rs

```diff
@@ -161,83 +177,126 @@ impl S3Storage {
-        if config.force_path_style {
-            client.config_mut().addressing_style = AddressingStyle::Path;
+        // static credentials are used with minio
+        if let Some(access_key_pair) = &config.access_key_pair {
+            let creds = Credentials::from_keys(
+                (*access_key_pair.access_key).to_owned(),
+                (*access_key_pair.secret_access_key).to_owned(),
+                access_key_pair
+                    .session_token
+                    .as_deref()
+                    .map(|s| s.to_owned()),
+            );
+            Self::maybe_assume_role(config, client, creds)
+        } else {
+            let creds = util::new_credentials_provider(client.clone());
+            Self::maybe_assume_role(config, client, creds)
         }
-        Ok(S3Storage { config, client })
     }
 
-    fn maybe_assume_role<P, D>(
+    fn maybe_assume_role<Creds, Http>(
         config: Config,
-        cred_provider: P,
-        dispatcher: D,
-    ) -> io::Result<S3Storage>
+        client: Http,
+        credentials_provider: Creds,
+    ) -> io::Result<Self>
     where
-        P: ProvideAwsCredentials + Send + Sync + 'static,
-        D: DispatchSignedRequest + Send + Sync + 'static,
+        Http: HttpClient + 'static,
+        Creds: ProvideCredentials + 'static,
     {
         if config.role_arn.is_some() {
-            // try use role arn anyway with current creds when it's not nil.
-            let bucket_region = none_to_empty(config.bucket.region.clone());
-            let bucket_endpoint = config.bucket.endpoint.clone();
-            let region = util::get_region(&bucket_region, &none_to_empty(bucket_endpoint))?;
-            // cannot use the same dispatcher because of move, so use another http client.
-            let sts = StsClient::new_with(util::new_http_client()?, cred_provider, region);
             let duration_since_epoch = SystemTime::now()
                 .duration_since(SystemTime::UNIX_EPOCH)
                 .unwrap();
             let timestamp_secs = duration_since_epoch.as_secs();
-            let cred_provider = StsAssumeRoleSessionCredentialsProvider::new(
-                sts,
-                String::clone(config.role_arn.as_deref().unwrap()),
-                format!("{}", timestamp_secs),
-                config.external_id.as_deref().cloned(),
-                // default duration is 15min
-                None,
-                None,
-                None,
-            );
-            Self::new_creds_dispatcher(config, dispatcher, cred_provider)
+
+            let mut builder = AssumeRoleProvider::builder(config.role_arn.as_deref().unwrap())
+                .session_name(format!("{}", timestamp_secs));
+
+            if let Some(external_id) = &config.external_id {
+                builder = builder.external_id(external_id.as_str());
+            }
+
+            if let Some(region) = &config.bucket.region {
```

Comment:
Who is responsible to clarify that?

Reply:
From aws document, https://docs.aws.amazon.com/STS/latest/APIReference/welcome.html
> AWS recommends using Regional AWS STS endpoints instead of the global endpoint to reduce latency, build in redundancy, and increase session token validity. 

It's better to use regional STS here. But

> Most AWS Regions enable operations in all AWS services by default. These Regions automatically activate for use with AWS STS. Some Regions, such as Asia Pacific (Hong Kong), must be manually enabled.

So, we should let user make sure the S3 bucket region has enabled STS service. Consider this make sense. It's okay to add this assumption. And we'd better test the behavior after this PR merged.